iree/compiler/Dialect/Util/Analysis/Attributes/Range.cpp

```diff
@@ -0,0 +1,254 @@
+// FpRangeValueElement
+//===----------------------------------------------------------------------===//
+
+const char FpRangeValueElement::ID = 0;
+
+// Returns whether the given type is a valid scalar fp type or a shaped type
+// of fp types.
+// TODO: getElementTypeOrSelf?
+static bool isFpType(Type type) {
+  return getElementTypeOrSelf(type).isa<FloatType>();
+}
+
+void FpRangeValueElement::initializeValue(Value value, DFX::Solver &solver) {
+  if (!isFpType(value.getType())) {
+    indicatePessimisticFixpoint();
+    return;
+  }
+}
+
+ChangeStatus FpRangeValueElement::updateValue(Value value,
+                                              DFX::Solver &solver) {
+  // TODO: This could be modularized substantially and made common with
+  // other attributes.
+  FpRangeState newState = getState();
+
+  // If this is a block argument to a supported parent operation, then
+  // remap the value we are working on to the correct input from the
+  // parent. This works because the statistics we are accumulating flow
+  // layer-wise, and we don't need to pay particular attention to specific
+  // loop structures.
+  // TODO: We shouldn't need to hard switch on LinalgOp here and should
+  // be relying on some kind of concept/interface. It just isn't
+  // clear what that would be.
+  if (auto valueBlockArg = value.dyn_cast<BlockArgument>()) {
+    Block *ownerBlock = valueBlockArg.getOwner();
+    if (auto linalgParent = llvm::dyn_cast_or_null<linalg::LinalgOp>(
+            ownerBlock->getParentOp())) {
+      value = linalgParent->getOperand(valueBlockArg.getArgNumber());
+      LLVM_DEBUG(dbgs() << "  ++ REMAP LINALG BLOCK ARG TO: " << value << "\n");
+    }
+  }
+
+  // If we can get a full potential value set, then we can derive from that.
+  auto pvs = solver.getElementFor<ConstantAttributePVS>(
+      *this, Position::forValue(value), DFX::Resolution::OPTIONAL);
+  if (pvs.isValidState() && !pvs.isUndefContained()) {
+    for (Attribute constValue : pvs.getAssumedSet()) {
+      if (auto scalarValue = constValue.dyn_cast<FloatAttr>()) {
+        FpRangeStats stats;
+        stats.reset();
+        stats.addDomainValue(scalarValue.getValueAsDouble());
+        newState.setAssumed(stats);
+        newState.indicateOptimisticFixpoint();
+      } else if (auto elements = constValue.dyn_cast<ElementsAttr>()) {
+        FpRangeStats stats;
+        stats.reset();
+        for (APFloat elementValue : elements.getValues<APFloat>()) {
+          stats.addDomainValue(elementValue.convertToDouble());
+        }
+        newState.setAssumed(stats);
+        LLVM_DEBUG(dbgs() << "*** COMPUTED KNOWN RANGE: " << stats.getAsStr()
+                          << "\n");
+        newState.indicateOptimisticFixpoint();
+      } else {
+        // Unknown.
+        // TODO
+        LLVM_DEBUG(dbgs() << "UNKNOWN ATTRIBUTE: " << constValue << "\n");
+        newState.indicatePessimisticFixpoint();
+      }
+    }
+  }
+
+  if (!newState.isAtFixpoint()) {
```

Comment:
nit: check if fixpoint and return here so that you can de-nest the below one level